### PR TITLE
Split into several libraries

### PR DESCRIPTION
**Pull request by falseresync**

### CODE REVIEW

File: cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package io.github.cottonmc.cotton;
+package io.github.cottonmc.cotton.datapack;
```
Comment (Juuxel) [inline cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java]: `CottonInitializer`s aren't really datapack-related, leaving it in the root package might be good. I don't mind having it in the datapack module, though.

File: src/main/resources/fabric.mod.json [hunk 2]
```diff
@@ -1,32 +1,13 @@
...
+  "version": "${version}",
   "name": "Cotton",
-  "icon": "assets/cotton/icon.png",
-  "description": "Common materials for inter-compatibility and ease of use.",
+  "icon": "icon.png",
+  "description": "Feature-rich set of libraries for the most important things you need",
   "licence": "MIT",
   "contact": {
-    "maven": "http://server.bbkr.space:8081/artifactory/libs-snapshot/io/github/cottonmc/cotton/",
-    "sources": "https://github.com/CottonMC/cotton",
```
Comment (Juuxel) [inline src/main/resources/fabric.mod.json]: Why did you remove the source link?

File: src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java [hunk 3]
```diff
@@ -1,118 +0,0 @@
-package io.github.cottonmc.cotton.test.util;
```
Comment (Juuxel) [inline src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java]: Was this file removed completely?

File: cotton-commons/src/main/java/io/github/cottonmc/cotton/commons/CottonCommons.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+public class CottonCommons implements ModInitializer {
+
+	public static final String MODID = "cotton-commons";
+
+	public static final ItemGroup commonGroup = FabricItemGroupBuilder.build(
+			new Identifier("cotton-commons", "common_tab"), () -> new ItemStack(Blocks.LIGHT_BLUE_WOOL));
+
+	public static final ModLogger logger = new ModLogger(MODID, "COTTON COMMONS");
+	public static final CottonCommonsConfig config = ConfigManager.loadConfig(CottonCommonsConfig.class);;
```
Comment (Juuxel) [inline cotton-commons/src/main/java/io/github/cottonmc/cotton/commons/CottonCommons.java]: These three fields should probably be all-caps, since they're `static final`.

Comment (Juuxel) [reply cotton-commons/src/main/java/io/github/cottonmc/cotton/commons/CottonCommons.java]: Also, there's a double semicolon there.

File: cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java (hunk 1 above)
Comment (falseresync) [reply cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java]: There’s no default module, so it is in datapack

File: src/main/resources/fabric.mod.json (hunk 2 above)
Comment (falseresync) [reply src/main/resources/fabric.mod.json]: I replaced it with homepage everywhere

File: src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java (hunk 3 above)
Comment (falseresync) [reply src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java]: Yup, since it is no use as far as I can see

Comment (Juuxel) [reply src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java]: It's a unit test class for `Identifiers`, so it shouldn't have any usages.

File: cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java (hunk 1 above)
Comment (Juuxel) [reply cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java]: Yep, I'm suggesting that it should go to the root package (`io.g.cmc.cotton`) in the datapack module.

Comment (falseresync) [reply cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java]: Would probably only cause confusion

Comment (Juuxel) [reply cotton-datapack/src/main/java/io/github/cottonmc/cotton/datapack/CottonInitializer.java]: True.

File: src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java (hunk 3 above)
Comment (falseresync) [reply src/test/java/io/github/cottonmc/cotton/test/util/TestIdentifiers.java]: ¯\_(ツ)_/¯

File: .gitignore [hunk 5]
```diff
@@ -1,32 +1,28 @@
...
+out
+classes
+
+# Naughty filesystems
 
-# Files from bad operating systems :^)
-Thumbs.db
-.DS_Store
-local.properties
 .directory
```
Comment (Boundarybreaker) [inline .gitignore]: Will this really deal with Thumbs.db, .DS_Store, and local.properties? I don't think it's really necessary to change the gitignore; it's been working fine so far.

File: src/main/resources/fabric.mod.json [hunk 6]
```diff
@@ -1,32 +1,13 @@
...
-  },
-  "mixins": [
-    "cotton.common.json"
-  ],
-  "requires": {
-    "fabricloader": ">=0.4.0",
-    "fabric": "*"
-  },
-  "custom": {
-    "modmenu:api": true
```
Comment (Boundarybreaker) [inline src/main/resources/fabric.mod.json]: We should still be marking this `modmenu:api`, shoudn't we? The fact that it's bundling the submods doesn't mean that it isn't an API.

File: modules/cotton-player-events/src/main/resources/fabric.mod.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    "issues": "https://github.com/CottonMC/cotton/issues"
+  },
+  "mixins": [
+    "mixins.cotton-player-events.json"
+  ],
+  "depends": {
+    "fabricloader": ">=0.4.0",
+    "fabric-api-base": "*"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-player-events/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-logging/src/main/resources/fabric.mod.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  "icon": "icon.png",
+  "licence": "MIT",
+  "contact": {
+    "homepage": "https://curseforge.com/minecraft/mc-mods/cotton",
+    "issues": "https://github.com/CottonMC/cotton/issues"
+  },
+  "depends": {
+    "fabricloader": ">=0.4.0"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-logging/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-datapack/src/main/resources/mixins.cotton-datapack.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    "MixinClientDataLoad",
+    "MixinCottonInitializerClient"
+  ],
+  "server": [
+    "MixinCottonInitializerServer"
+  ],
+  "injectors": {
+    "defaultRequire": 1
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-datapack/src/main/resources/mixins.cotton-datapack.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-datapack/src/main/resources/fabric.mod.json [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    "mixins.cotton-datapack.json"
+  ],
+  "depends": {
+    "fabricloader": ">=0.4.0",
+    "cotton-logging": "*",
+    "cotton-config": "*",
+    "jankson": "*",
+    "fabric-commands-v0": "*"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-datapack/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-config/src/main/resources/fabric.mod.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  "contact": {
+    "homepage": "https://curseforge.com/minecraft/mc-mods/cotton",
+    "issues": "https://github.com/CottonMC/cotton/issues"
+  },
+  "depends": {
+    "fabricloader": ">=0.4.0",
+    "cotton-logging": "*",
+    "jankson": "*"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-config/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-commons/src/main/resources/fabric.mod.json [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    "fabricloader": ">=0.4.0",
+    "cotton-logging": "*",
+    "cotton-config": "*",
+    "cotton-datapack": "*",
+    "jankson": "*",
+    "fabric-api-base": "*",
+    "fabric-item-groups-v0": "*",
+    "fabric-tag-extensions-v0": "*"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-commons/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-cauldron/src/main/resources/mixins.cotton-cauldron.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+  "package": "io.github.cottonmc.cotton.cauldron.mixins",
+  "compatibilityLevel": "JAVA_8",
+  "refmap": "mixins.cotton-cauldron.refmap.json",
+  "mixins": [
+    "MixinCauldronBehavior"
+  ],
+  "injectors": {
+    "defaultRequire": 1
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-cauldron/src/main/resources/mixins.cotton-cauldron.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: modules/cotton-cauldron/src/main/resources/fabric.mod.json [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  },
+  "mixins": [
+    "mixins.cotton-cauldron.json"
+  ],
+  "depends": {
+    "fabricloader": ">=0.4.0",
+    "cotton-commons": "*",
+    "libcd": "*"
+  }
+}
```
Comment (Boundarybreaker) [inline modules/cotton-cauldron/src/main/resources/fabric.mod.json]: Needs to be marked as `modmenu:api` and have the parent marked as cotton

File: .gitignore (hunk 5 above)
Comment (falseresync) [reply .gitignore]: Dunno, that’s an automatic thing. I don’t thing it worths deleting

Comment (Boundarybreaker) [reply .gitignore]: I've run into issues in the past where these, if not gitignore'd, will appear in source when you don't want them to. I don't think we should have any gitignore changes.

File: src/main/resources/fabric.mod.json (hunk 6 above)
Comment (falseresync) [reply src/main/resources/fabric.mod.json]: Yup, completely forgot about this one

File: .gitignore (hunk 5 above)
Comment (falseresync) [reply .gitignore]: Ah, right, they’re deleted, not added

File: build.gradle [hunk 15]
```diff
@@ -1,151 +1,119 @@
...
+    modApi ("io.github.cottonmc:Jankson:1.0.0+j1.1.2") { transitive = false }
+    modApi ("io.github.cottonmc:LibCD:1.3.0+1.14.3") { transitive = false }
+    include "io.github.cottonmc:Jankson:1.0.0+j1.1.2"
+    include "io.github.cottonmc:LibCD:1.3.0+1.14.3"
+
+    compileOnly ("com.google.code.findbugs:jsr305:3.0.2") { transitive = false }
+
+    // For ModMenu compatibility tests
+    modImplementation "net.fabricmc.fabric-api:fabric-api:0.3.0+build.187"
+    modImplementation "io.github.prospector.modmenu:ModMenu:1.6.3-94"
```
Comment (Juuxel) [inline build.gradle]: This (or both, depending on if Fabric API is needed during compilation) can be `modRuntime`.

Comment (falseresync) [reply build.gradle]: Nope, they can't. Tested

File: .gitignore (hunk 5 above)
Comment (falkreon) [reply .gitignore]: yeah, definitely don't remove these, we have a lot of different systems creating a lot of hidden files

but local.properties should move down into the gradle private files section
